soci/soci_convert.go

```diff
@@ -0,0 +1,317 @@
+			return nil, err
+		}
+	}
+
+	return &ociIndexDesc, nil
+}
+
+// buildSociIndexesv2ForPlatforms builds a SOCI index for each specified platform
+func (b *IndexBuilder) buildSociIndexesv2ForPlatforms(ctx context.Context, img images.Image, platforms []ocispec.Platform) ([]*IndexWithMetadata, error) {
+	var indexes []*IndexWithMetadata
+	for _, platform := range platforms {
+		index, err := b.Build(ctx, img,
+			WithPlatform(platform),
+			// Don't set a GC label on the SOCI indexes because their GC will be attached to
+			// the converted image's OCI index. If we label them, they will not get
+			// gcd with the rest of the converted image
+			WithNoGarbageCollectionLabel(),
+			withIndexVersion(V2),
+		)
+		if err != nil {
+			// If a platform produces an empty index, try other platforms.
+			// This is unlikely, but could happen if one platform doesn't have
+			// any layers larger than min layer size. We should still try to index
+			// other platforms.
+			if errors.Is(err, ErrEmptyIndex) {
+				continue
+			}
+			return nil, err
+		}
+		indexes = append(indexes, index)
+	}
+
+	// If we didn't produce any indexes, return that as an error
+	if len(indexes) == 0 {
+		return nil, ErrEmptyIndex
+	}
+	return indexes, nil
+}
+
+// newOciIndex creates an OCI index object for the converted image.
+// If `img.Target` is already an OCI index (or docker manifest list), newOciIndex loads that index,
+// otherwise, it creates a new OCI index containing just the single image target.
+func (b *IndexBuilder) newOciIndex(ctx context.Context, img images.Image) (ocispec.Index, error) {
+	if images.IsIndexType(img.Target.MediaType) {
+		// Load the original target
+		ra, err := b.contentStore.ReaderAt(ctx, img.Target)
+		if err != nil {
+			return ocispec.Index{}, err
+		}
+		b, err := io.ReadAll(io.NewSectionReader(ra, 0, ra.Size()))
+		if err != nil {
+			return ocispec.Index{}, err
+		}
+		err = ociutil.ValidateMediaType(b, img.Target.MediaType)
+		if err != nil {
+			return ocispec.Index{}, err
+		}
+
+		var ociIndex ocispec.Index
+		err = json.Unmarshal(b, &ociIndex)
+		// Some Registries don't like it when you push a Docker v2 manifest
+		// To increase compatibility, change the media type to OCI image manifest
+		ociIndex.MediaType = ocispec.MediaTypeImageIndex
+		return ociIndex, err
+	}
+	return ocispec.Index{
+		Versioned: specs.Versioned{
+			SchemaVersion: 2,
+		},
+		MediaType: ocispec.MediaTypeImageIndex,
+		Manifests: []ocispec.Descriptor{
+			img.Target,
+		},
+	}, nil
+}
+
+// annotateImages adds the SOCI index digest to the corresponding image manifest descriptor, modifying `ociIndex` in the process.
+// Adding annotations modifies the image manifest, so annotateImages also pushes the modified image manifests to the blobStore,
+// computes the new image digests, and modified the image descriptors in `ociIndex`
+func (b *IndexBuilder) annotateImages(ctx context.Context, ociIndex *ocispec.Index, sociIndexes []*IndexWithMetadata) error {
```

Comment:
Would it be better to just remove the modified index, instead of modifying it in-place? Might make it a little clearer what exactly is being modified

Reply:
Do you mean copy?

I initially wanted all of these functions to return a new index, but indexes have a lot of pointers and arrays and maps, so you would need to be sure to do deep copies to avoid accidentally keeping references. I figured just doing the modifications in place means we don't have to worry about accidentally keeping references because you can't assume the original is unmodified.

Reply:
Yeah I kind of see what you mean, I was thinking that since this only modifies the manifests you can have the func take a list of manifests, clone just the manifests list, modify the new list, then return just the manifests, but I can see that being a little clunky to work with unless done carefully.

My main "problem" here is that, aside from the comment, it's not immediately obvious that this modifies the index, but it does make sense in this context. I'm not super opinionated on this, though.

Reply:
https://github.com/awslabs/soci-snapshotter/issues/1536